File: day2/dive.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"))
  .toString("utf-8")
  .split("\n")
  .map((line) =>
    line.split(" ").map((v, i) => (i === 1 ? parseInt(v, 10) : v))
  );

// Part 1
const forward = input
  .filter((item) => item[0] === "forward")
  .reduce((acc, item) => acc + item[1], 0);
const up = input
  .filter((item) => item[0] === "up")
  .reduce((acc, item) => acc + item[1], 0);
const down = input
  .filter((item) => item[0] === "down")
  .reduce((acc, item) => acc + item[1], 0);

console.log(forward, down - up, forward * (down - up));

// Part 2
const { aim, position, depth } = input.reduce(
  (acc, item) => {
    const [command, value] = item;
    const { aim, position, depth } = acc;
    switch (command) {
      case "forward":
        return { aim, position: position + value, depth: depth + value * aim };
      case "down":
        return { aim: aim + value, position, depth };
      case "up":
        return { aim: aim - value, position, depth };
    }
    return acc;
  },
  {
    aim: 0,
    position: 0,
    depth: 0,
  }
);
console.log(aim, position, depth, position * depth);

File: day2/Dive.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

record Command(String action, Integer value) {
    public static Command fromLineString(String line) {
        var parts = line.split(" ");
        return new Command(parts[0], Integer.parseInt(parts[1], 10));
    }
}
record Result(Integer aim, Integer position, Integer depth) {
    public static Result add(Result result, Command command) {
        return switch (command.action()) {
            case "forward" -> new Result(result.aim, result.position + command.value(), result.depth + (result.aim * command.value()));
            case "up" -> new Result(result.aim - command.value(),result.position,result.depth);
            case "down" -> new Result(result.aim + command.value(), result.position,result.depth);
            default -> result;
        };
    }
    public static Result sum(Result a, Result b) {
        return new Result(a.aim + b.aim, a.position + b.position, a.depth + b.depth);
    }
}

class Dive {
    public static void main(String[] args) throws IOException {
        var lines = Files.lines(Paths.get("day2/input.txt")).map(Command::fromLineString);
        var result = lines.reduce(new Result(0,0,0), Result::add, Result::sum);
        System.out.printf("Part1: %d, Part2: %d", result.aim() * result.position(), result.position() * result.depth());
    }
}